src/lib/optimizer/strategy/between_composition_rule.cpp

```diff
@@ -0,0 +1,219 @@
+      if (lower_bound_value_expression != nullptr && upper_bound_value_expression != nullptr) {
+        const auto between_node = PredicateNode::make(
+            std::make_shared<BetweenExpression>(current_column_expression, lower_bound_value_expression,
+                                                upper_bound_value_expression, left_inclusive, right_inclusive));
+        between_nodes.push_back(between_node);
+      } else {
+        predicate_nodes.insert(std::end(predicate_nodes), std::begin(node_scope), std::end(node_scope));
+      }
+      upper_bound_value_expression = nullptr;
+      lower_bound_value_expression = nullptr;
+      left_inclusive = true;
+      right_inclusive = true;
+      node_scope = std::vector<std::shared_ptr<AbstractLQPNode>>();
+      current_column_expression = boundary.column_expression;
+    }
+
+    if (boundary.type == ColumnBoundaryType::UpperBoundaryInclusive) {
+      if (upper_bound_value_expression == nullptr ||
+          upper_bound_value_expression->value > boundary.value_expression->value) {
+        upper_bound_value_expression = boundary.value_expression;
+        right_inclusive = true;
+        node_scope.push_back(boundary.node);
+      }
+    } else if (boundary.type == ColumnBoundaryType::LowerBoundaryInclusive) {
+      if (lower_bound_value_expression == nullptr ||
+          lower_bound_value_expression->value < boundary.value_expression->value) {
+        lower_bound_value_expression = boundary.value_expression;
+        left_inclusive = true;
+        node_scope.push_back(boundary.node);
+      }
+    } else if (boundary.type == ColumnBoundaryType::UpperBoundaryExclusive) {
+      if (upper_bound_value_expression == nullptr ||
+          upper_bound_value_expression->value <= boundary.value_expression->value) {
+        upper_bound_value_expression = boundary.value_expression;
+        right_inclusive = false;
+        node_scope.push_back(boundary.node);
+      }
+    } else if (boundary.type == ColumnBoundaryType::LowerBoundaryExclusive) {
+      if (lower_bound_value_expression == nullptr ||
+          lower_bound_value_expression->value <= boundary.value_expression->value) {
+        lower_bound_value_expression = boundary.value_expression;
+        left_inclusive = false;
+        node_scope.push_back(boundary.node);
+      }
+    }
+  }
+
+  // Apply boundary check again for the last column
+  if (lower_bound_value_expression != nullptr && upper_bound_value_expression != nullptr) {
+    const auto between_node = PredicateNode::make(
+        std::make_shared<BetweenExpression>(current_column_expression, lower_bound_value_expression,
+                                            upper_bound_value_expression, left_inclusive, right_inclusive));
+    between_nodes.push_back(between_node);
+  } else {
+    predicate_nodes.insert(std::end(predicate_nodes), std::cbegin(node_scope), std::cend(node_scope));
```

Comment:
From our understanding, `node_scope` contains all nodes that were at some point an upper or lower bound for an between expression.
It does make sense to track them, as we might not find a corresponding lower/upper bound and need to re-insert them. However, don't we have to remove them from `node_scope` as soon as we have both a lower and an upper bound? From our understanding, this should lead to additional nodes in the LQP, causing all tests for this rule to fail.
We could not build the tests (we are getting the same compiler error as the [ci-server](https://hyrise-ci.epic-hpi.de/blue/organizations/jenkins/hyrise/hyrise/detail/PR-1481/1/pipeline)), so we are not perfectly sure about it.
Maybe there is there something we missed or misunderstood?